server/controller/src/main/java/ai/starwhale/mlops/datastore/WalManager.java

```diff
@@ -0,0 +1,350 @@
+                try {
+                    //noinspection BusyWait
+                    Thread.sleep(1000);
+                } catch (InterruptedException ex) {
+                    log.warn("interrupted", e);
+                }
+            }
+        }
+    }
+
+    private enum PopulationStatus {
+        TERMINATED,
+        BUFFER_FULL,
+        NO_MORE_ENTRIES
+    }
+
+    private PopulationStatus populateOutput() {
+        synchronized (this.entries) {
+            while (!this.terminated && this.entries.isEmpty()) {
+                try {
+                    this.entries.wait(this.walWaitIntervalMillis);
+                } catch (InterruptedException e) {
+                    log.warn("interrupted", e);
+                }
+            }
+            if (this.terminated && this.entries.isEmpty()) {
+                return PopulationStatus.TERMINATED;
+            }
+        }
+        for (; ; ) {
+            Wal.WalEntry entry;
+            synchronized (this.entries) {
+                if (this.entries.isEmpty()) {
+                    break;
+                }
+                entry = this.entries.getFirst();
+            }
+            if (CodedOutputStream.computeMessageSizeNoTag(entry) + this.outputStream.getOffset()
+                    > this.walMaxFileSize) {
+                if (this.outputStream.getOffset() == 0) {
+                    // huge single entry
+                    log.error(
+                            "data loss: discard unexpected huge entry. size={} table={} schema={} records count={}",
+                            entry.getSerializedSize(),
+                            entry.getTableName(),
+                            entry.getTableSchema(),
+                            entry.getRecordsCount());
+                } else {
+                    return PopulationStatus.BUFFER_FULL;
+                }
+            } else {
+                try {
+                    entry.writeDelimitedTo(this.outputStream);
+                } catch (IOException e) {
+                    log.error("data loss: unexpected exception", e);
+                }
+            }
+            synchronized (this.entries) {
+                this.entries.removeFirst();
+            }
+        }
+        if (this.outputStream.getOffset() >= this.walFileSize) {
+            return PopulationStatus.BUFFER_FULL;
+        }
+        return PopulationStatus.NO_MORE_ENTRIES;
+    }
+
+    private void writeToObjectStore(boolean clearOutput) {
+        int compressedSize;
+        try {
+            var inBuf = this.outputBuffer.asByteBuffer();
+            var outBuf = this.compressedBuffer.asByteBuffer();
+            if (inBuf.hasArray()) {
+                compressedSize = Snappy.compress(inBuf.array(), 0, this.outputStream.getOffset(), outBuf.array(), 0);
+            } else {
+                inBuf.limit(this.outputStream.getOffset());
+                compressedSize = Snappy.compress(inBuf, outBuf);
```

Comment:
Data may be lost if the compressed data size is larger than the original data size.